tests/robustness/client/client.go

```diff
@@ -371,3 +372,56 @@ func toWatchEvent(event clientv3.Event) (watch model.WatchEvent) {
 	}
 	return watch
 }
+
+type ClientSet struct {
+	idProvider identity.Provider
+	baseTime   time.Time
+
+	closed  bool
+	clients []*RecordingClient
+	reports []report.ClientReport
+}
+
+func NewSet(ids identity.Provider, baseTime time.Time) *ClientSet {
+	return &ClientSet{
+		idProvider: ids,
+		baseTime:   baseTime,
+
+		clients: []*RecordingClient{},
+	}
+}
+
+func (cs *ClientSet) NewClient(endpoints []string) (*RecordingClient, error) {
+	if cs.closed {
+		return nil, errors.New("The ClientSet is already closed")
+	}
+	cli, err := NewRecordingClient(endpoints, cs.idProvider, cs.baseTime)
+	if err != nil {
+		return nil, err
+	}
+	cs.clients = append(cs.clients, cli)
+	return cli, nil
+}
+
+func (cs *ClientSet) Reports() []report.ClientReport {
+	if !cs.closed {
+		cs.Close()
+	}
+	return cs.reports
+}
+
+func (cs *ClientSet) Close() {
+	if cs.closed {
+		return
+	}
+	cs.closed = true
+	cs.reports = make([]report.ClientReport, 0, len(cs.clients))
```

Comment:
It is incorrect to calculate reports only on `Close` as it's always called after `Reports`. Reports is called within function while Close is deferred. Solution. Close during Reports call.

Reply:
This should make the code cleared because it would keep Closing focused on closing, while Report on generating report.

Reply:
Done.